### PR TITLE
Using hawkular-metrics scenario

### PR DESCRIPTION
**Pull request by e-minguez**
# Introduction
This PR will enable users to have a new hawkular metrics scenario in katacoda to explain:

* What metrics look like (all metrics pods and their roles)
* How to see real metrics (both web gui and with oc adm top)
* What an HPA is
* How to use an HPA

### Image
I've added a new `openshift-36-metrics` image with a new script `metrics.sh` that deploys metrics at build time.

### Scripts
The current katacoda process requires the OpenShift environment to be clean (that means, no `/openshift.*/` folders) as it is generated at boot time with [some magic](https://github.com/openshift-evangelists/intro-katacoda/issues/72#issuecomment-342909904)

This includes the `etcd` database therefore all objects should be wiped in the image (including metrics stuff...), that's why even if I run the installation at build time in the metrics.sh script, it is required to be executed at runtime again to be able to create all dcs, rcs, etc.

I didn't removed the deployment at build time to ensure all the requisites are included in the image.

### Markdown files
The scenario should be completed using the markdown files provided that includes some screenshots and procedures to observe metrics (both web interface and command line) and create and use `hpa` objects.

## Current status
Building the local image with packer works fine (see this [PR](https://github.com/openshift-evangelists/intro-katacoda/pull/107) for a required fix on building the image on Fedora).
The image requires 2 GB of RAM to run otherwise the metrics pods won't be able to start.
I've been able to test it with the openshift-middleware image and deploying the metrics manually following the script instructions.

## Questions
* Can I test somehow the image in a real katacoda environment? I've been struggling with my katacoda instance but I cannot see how to add a new image to the environment.
* I think the deletion of the local data is mandatory in order to generate the proper certificates, hostnames, etc. am I right? In that case, the ansible playbook to deploy metrics should be executed at runtime (it is in this PR)

### CODE REVIEW

**Comment by e-minguez** [comment]
I've uploaded the scenario content.

**Comment by e-minguez** [comment]
So, following the instructions provided in the README file, I'm pinging @mhausenblas (I don't have access to that trello board, but I created [a card](https://trello.com/c/Gxw6cLpb/1000-5-katacoda-new-scenarios) in my team's board) for a review on this scenario.

**Comment by mhausenblas** [comment]
@e-minguez I've sent you an invite to our Trello board via mail now, pls let me know if you're having issues getting in.

**Comment by jorgemoralespou** [comment]
@e-minguez what's the start time of the VM since you need to run the metrics.sh script?

@BenHall @mhausenblas I see more and more evidences that we should maybe move into an "oc cluster up" startup process, since we will avoid this additional complexity of bootstrapping clusters for additional capabilities that we're no experts on.

Else, we can just use the ansible installer, like @e-minguez does, and revisit why the etcd is wiped and sort that out. Definitely, as we create more images for different scenarios, things will get more complicated. 

And then will come the problem of updating it to a new release of openshift. Will this image and this scripts be maintained by @e-minguez forever. We need to use one of the supported install processes, and those are ansible or oc cluster up.

**Comment by e-minguez** [comment]
It shouldn't take too long as the images are there and all the prerrequisites are accomplished at build time (that's why the installation playbook is executed twice). Playbook takes about 1 minute https://imgur.com/a/tBcdo on the current openshift-middleware image, I think using the openshift-metrics image will reduce it

**Comment by jorgemoralespou** [comment]
@e-minguez Does that mean that when starting an scenario, the user has to wait 1 minute before he can proceed?
Is the openshift cluster already up? and if so, can the user go to the metrics console to accept the certificate? 
If the provisioning has not finished, what will be the experience?
Is metrics using the same certificate as katacoda, so the certificate does not need to be accepted by the user?

**Comment by e-minguez** [comment]
Yes, the metrics provisioning takes ~1 minute even if the images are cached yet. 
The OCP cluster is ready and it is not required to accept any certificate as it is the katacoda certificate.

**Comment by jorgemoralespou** [comment]
@e-minguez but if a user logs into the console and metrics are not yet deployed? What will he see?
If then, metrics deployment finishes, will he need to reload the page, or will metrics just show up?

**Comment by mhausenblas** [comment]
I think that while it's not stelar UX it's manageable if things are slower to start up (we have that in other scenarios as well) as long as we do proper expectation management. So, as long it's called out upfront (ideally in bold) that it takes a bit, for example `note that it may take up to a minute until you see XXX`, it's fine.

**Comment by e-minguez** [comment]
Jorge's concerns:
* Explain you need to be logged as `system:admin` to see metrics pods -> done
* Timing -> Added warning note for downscale[1]

Michael's concerns:
* Added a note in the `intro.md` file with a warning and as a bonus a link to the official documentation so while the user is waiting he can read about metrics :)

More stuff:
* Added kubernetes/guestbook to the image to speed up app creation
* Reduced CPU hog operation to 60 seconds (just what the hpa needs)
* Added more clarifications

[1] Jorge told me about `horizontal-pod-autoscaler-downscale-delay` controller parameters, but in OCP 3.6 it is hardcoded and it doesn't work 
OCP hpa delay
So I've added a warning to the user that it can take up to 5 minutes for the downscale operation. There is nothing documented in our official docs, so I'll open a bugzilla to include this. Thanks Jorge!

cc @jorgemoralespou @mhausenblas

**Comment by e-minguez** [comment]
BZ opened -> https://bugzilla.redhat.com/show_bug.cgi?id=1532584

**Comment by e-minguez** [comment]
Is there anything else you require to have this merged? Thanks!

**Comment by mhausenblas** [comment]
No @e-minguez everything is fine. Apologies for the delay, let me do a final walk-trough and then I merge. At which URL is the scenario currently available?

**Comment by e-minguez** [comment]
There is no scenario available yet. It requires the new image to be built, so I've been testing it using [this](https://katacoda.com/eminguez/courses/intro-openshift/deploy-metrics) and running the following (which it will be the "startup" script:
```
ansible-playbook -i myinventory openshift-ansible-openshift-ansible-${OAVERSION}/playbooks/byo/openshift-cluster/openshift-metrics.yml -e "openshift_metrics_install_metrics=True" -e "openshift_metrics_hawkular_hostname=hawkular-metrics-$(awk '/subdomain/ { print $2 }' /etc/origin/master/master-config.yaml | sed 's/-80-/-443-/g')" > /dev/null 2>&1

systemctl restart origin.service
rm -Rf myinventory openshift-ansible-openshift-ansible-${OAVERSION}/
```

The markdown docs are the ones in this PR.

I think it would be nice to create the image in katacoda, then enable the scenario "hidden" from the list to be able to test it "real".

**Comment by mhausenblas** [comment]
Yes, thanks, https://katacoda.com/eminguez/courses/intro-openshift/deploy-metrics is what I was looking for

**Comment by GrahamDumpleton** [comment]
This has to go under a category now so needs to be changed to new structure.

Should there be a new category called 'infrastructure' if it is demonstrating installing bits which are only done by administrators. It should not be under 'introduction', nor is 'middleware' appropriate.

**Comment by e-minguez** [comment]
It shows hpa features, the metrics bits are yet deployed for the user. Which category should fit into then?

**Comment by GrahamDumpleton** [comment]
Okay, misunderstand what this was for, I thought you were showing how to setup metrics given our existing katacoda vm images don't have them.

Either way, eventually we are going to want a category which isn't 'introduction', but more for 'advanced'. I don't want to see for example stuff like a/b or blue/green deployments in introduction which is really target are the core fundamentals. The horizontal pod autoscaler to me is on the edge of being a more advanced feature as it isn't something which a basic user is going to use.

Ultimately up to Michael, but don't want to putting advanced stuff in 'introduction'.

**Comment by mhausenblas** [comment]
Thanks for pointing this out @GrahamDumpleton and agreed, should not be under intro category. On the other hand *if* we go for a new category, for example `advanced` and we only have one scenario in there this doesn't look too good either. Can we (for now) do it in a way that it's not listed in the categories, IIRC Graham had something like this as well previously (kinda hidden). Thoughts?

**Comment by e-minguez** [comment]
Maybe creating `others` category to host this and when you have some more advanced scenarios create the `advance` category and move it back to it (or keep it in `others`)

My 2 cents.

**Comment by BenHall** [comment]
How about Metrics until we have additional scenarios and rename it to advance? After renaming would just need to add a redirect so users don't get lost. 

In terms of the environment, it will be done by the end of the week.

**Comment by e-minguez** [comment]
Hi folks,

Anything about this? Thanks!

**Comment by mhausenblas** [comment]
@e-minguez AFAICT the only thing blocking is moving it to a new category `Metrics` as suggested by @BenHall … or am I missing something?

**Comment by thesteve0** [comment]
Metrics seems like WAY too narrow of category. How about put it in intro,
like we did with FIS until we actually had enough for a middleware category.

**Comment by mhausenblas** [comment]
OK, since this one has priority I agree that (for now) we can put it in intro but I will make a note that we revisit the scenarios and come up with a new category for this and other ones that are not pure intro-level. Once @e-minguez confirms it's available in the intro category I merge it.

**Comment by e-minguez** [comment]
[It is](https://github.com/openshift-evangelists/intro-katacoda/pull/109/files#diff-bf83aaa1d8e209881d7d8e31bdf2ea49R40).
Regarding the merge, I'm not sure if the new metrics image should be build before the scenario go live or if can be merged but "hidden" until we verify everything works.

**Comment by mhausenblas** [comment]
Thanks! Not sure, need @BenHall to advice us here.

**Comment by BenHall** [comment]
Adding the scenario to the pathway makes it live to the public (If I'm following the question correctly)

**Comment by e-minguez** [comment]
There's two questions:

* Is the image ready to be used or should we create it before the scenario goes live?
* Can we make the scenario live but hidden then after testing it, making it public?

Thanks.

**Comment by BenHall** [comment]
1) We should wait. Sorry for the delay on this. I'm in the middle of delivering training and also working with the roadshow teams on their first events. 

2) Yes, you can merge the scenario. The only items listed on the site are defined in the pathway files

**Comment by e-minguez** [comment]
Ok then:

* I'll modify the pathway file to remove the metrics scenario to be able to:

1. Create the image without messing the scenario
2. Being able to test it after the image is ready without being live

After it is tested, I'll create another PR to make it live.

Is that ok? How should I access the scenario once it is live?

Thanks.

**Comment by mhausenblas** [comment]
That works for me, yes @e-minguez

**Comment by e-minguez** [comment]
I've just removed the scenario from the introduction pathway to be able to build the image and test it first.
Thanks @mhausenblas

**Comment by mhausenblas** [comment]
@e-minguez just checking in here to make sure I'm not missing anything and ensure that I'm not the blocker ;)

So, you're currently testing and let me know when I can merge, correct?

**Comment by e-minguez** [comment]
I did test it locally but in order to have the scenario, the image needs to be built so this needs to be merged. Once merged and @benhall create the image, it can be tested in katacoda then push it to production.
tl;dr yes, please merge it 😁

**Comment by e-minguez** [comment]
I did test it locally but in order to have the scenario, the image needs to be built so this needs to be merged. Once merged and @benhall create the image, it can be tested in katacoda then push it to production.
tl;dr yes, please merge it 😁

**Comment by e-minguez** [comment]
I've opened [an issue](https://github.com/openshift-evangelists/intro-katacoda/issues/121) to track the image build.
Thanks @mhausenblas !

**Comment by BenHall** [comment]
Not sure if this is an issue

```
    openshift-36-metrics: RUNNING HANDLER [openshift_metrics : restart master] ***************************
    openshift-36-metrics: fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check '(openshift.master.ha is not defined or not openshift.master.ha | bool) and (not (master_service_status_changed | default(false) | bool))' failed. The error was: error while evaluating conditional ((openshift.master.ha is not defined or not openshift.master.ha | bool) and (not (master_service_status_changed | default(false) | bool))): 'dict object' has no attribute 'master'\n\nThe error appears to have been in '/root/openshift-ansible-openshift-ansible-3.6.173.0.63-1/roles/openshift_metrics/handlers/main.yml': line 2, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n---\n- name: restart master\n  ^ here\n"}
    openshift-36-metrics:
    openshift-36-metrics: RUNNING HANDLER [openshift_metrics : restart master api] ***********************
    openshift-36-metrics:
    openshift-36-metrics: RUNNING HANDLER [openshift_metrics : restart master controllers] ***************
    openshift-36-metrics: 	to retry, use: --limit @/root/openshift-ansible-openshift-ansible-3.6.173.0.63-1/playbooks/byo/openshift-cluster/openshift-metrics.retry
    openshift-36-metrics:
    openshift-36-metrics: PLAY RECAP *********************************************************************
    openshift-36-metrics: localhost                  : ok=177  changed=44   unreachable=0    failed=1
```
